programs/castle-vault/src/instructions/rebalance.rs

```diff
@@ -60,55 +65,119 @@ pub struct Rebalance<'info> {
     //#[soteria(ignore)]
     pub port_reserve: AccountInfo<'info>,
 
+    /// CHECK: safe
+    //#[soteria(ignore)]
+    pub port_additional_states: AccountInfo<'info>,
+
+    /// CHECK: safe
+    //#[soteria(ignore)]
+    pub port_staking_pool: AccountInfo<'info>,
+
+    /// CHECK: safe
+    //#[soteria(ignore)]
+    pub port_reward_token_oracle: AccountInfo<'info>,
+
     pub clock: Sysvar<'info, Clock>,
 }
 
-impl TryFrom<&Rebalance<'_>> for AssetContainer<Reserves> {
-    type Error = Error;
-    fn try_from(r: &Rebalance<'_>) -> Result<AssetContainer<Reserves>> {
-        let flags: YieldSourceFlags = r.vault.get_yield_source_flags();
-
-        // NOTE: I tried pretty hard to get rid of these clones and only use the references.
-        // The problem is that these references originate from a deref() (or as_ref())
-        // and end up sharing lifetimes with the Context<Rebalance>.accounts lifetime,
-        // which means that the lifetimes are shared, preventing any other borrows
-        // (in particular the mutable borrow required at the end to save state)
-
-        // TODO is there a way to eliminate duplicate code here?
-        let solend = flags
-            .contains(YieldSourceFlags::SOLEND)
-            .as_option()
-            .map(|()| {
-                r.solend_reserve.key.eq(&r.vault.solend_reserve).as_result(
-                    Ok::<_, Error>(Reserves::Solend(Box::new(
-                        Account::<SolendReserve>::try_from(&r.solend_reserve)?
-                            .deref()
-                            .clone(),
-                    ))),
-                    ErrorCode::InvalidAccount,
-                )?
-            })
-            .transpose()?;
-
-        let port = flags
-            .contains(YieldSourceFlags::PORT)
-            .as_option()
-            .map(|()| {
-                r.port_reserve.key.eq(&r.vault.port_reserve).as_result(
-                    Ok::<_, Error>(Reserves::Port(Box::new(
-                        Account::<PortReserve>::try_from(&r.port_reserve)?
-                            .deref()
-                            .clone(),
-                    ))),
-                    ErrorCode::InvalidAccount,
-                )?
-            })
-            .transpose()?;
-
-        Ok(AssetContainer {
-            inner: [solend, port],
+fn assets_from_accounts(
```

Comment:
Why change away from implementing `TryFrom`?

Reply:
I decided to refactored the `try_from` a bit, we are back on `TryFrom` again now.